### ISSUE TITLE
Testing methods reject hex-encoded parameters

### ISSUE THREAD

**Issue by wolflo** [nomiclabs/hardhat#1732]
Testing methods like `evm_mine` and `evm_increaseTime` expect integer parameters. Unlike other methods, they reject these parameters if they are not decimal-encoded.

### To reproduce

`npx hardhat --version` -> `2.5.0`

Run `npx hardhat node` and make the following rpc calls:
```
> curl -H "Content-Type: application/json" -X POST --data \
        '{"id":1337,"jsonrpc":"2.0","method":"evm_increaseTime","params":[32]}' \
        http://localhost:8545
{"jsonrpc":"2.0","id":1337,"result":"41"}

> curl -H "Content-Type: application/json" -X POST --data \
        '{"id":1337,"jsonrpc":"2.0","method":"evm_increaseTime","params":["0x20"]}' \
        http://localhost:8545
{"jsonrpc":"2.0","id":1337,"error":{"code":-32602,"message":"Errors encountered in param 0: Invalid value \"0x20\" supplied to : number"}}
```

**Comment by wolflo** [comment]
This appears to be enforced in the test suite for `evm_increaseTime`. 

https://github.com/nomiclabs/hardhat/blob/c9e8a609be8b7751ce375a22159d4d2af162dad2/packages/hardhat-core/test/internal/hardhat-network/provider/modules/evm.ts#L127-L131

I would disagree that this should be the case, though. 

I don't know the reasoning for returning the total adjusted time from this method as a decimal-encoded value rather than hex-encoded, but that does seem to be intentional. However, ganache does still accept a hex-encoded argument to the method. Changing this behavior requires a special case for `evm_increaseTime` in any library that automatically encodes arguments to rpc methods as hex strings.

**Comment by fvictorio** [comment]
I think it makes sense to be more permissive and accept hex-encoded arguments in these methods. But we can't change the output of the method since that would be a breaking change.

@alcuadrado thoughts?

**Comment by wolflo** [comment]
Seems that the opposite is true, too. Ganache accepts decimal-encoded values for non-test methods like `eth_getBlockByNumber`, while hardhat (and geth) reject them.

`ganache-cli`
```
~ ❯ curl -H "Content-Type: application/json" -X POST --data \
        '{"id":1337,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":[1, true]}' \
        http://localhost:8545
{"id":1337,"jsonrpc":"2.0","result":null}

~ ❯ curl -H "Content-Type: application/json" -X POST --data \
        '{"id":1337,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x1", true]}' \
        http://localhost:8545
{"id":1337,"jsonrpc":"2.0","result":null}
```

`npx hardhat node`
```
~ ❯ curl -H "Content-Type: application/json" -X POST --data \
        '{"id":1337,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":[1, true]}' \
        http://localhost:8545
{"jsonrpc":"2.0","id":1337,"error":{"code":-32602,"message":"Errors encountered in param 0: Invalid value 1 supplied to : (QUANTITY | \"earliest\" | \"latest\" | \"pending\")/0: QUANTITY, Invalid value 1 supplied to : (QUANTITY | \"earliest\" | \"latest\" | \"pending\")/1: \"earliest\" | \"latest\" | \"pending\""}}

~ ❯ curl -H "Content-Type: application/json" -X POST --data \
        '{"id":1337,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x1", true]}' \
        http://localhost:8545
{"jsonrpc":"2.0","id":1337,"result":null}%
```

`geth --dev --http`
```
~ ❯ curl -H "Content-Type: application/json" -X POST --data \
        '{"id":1337,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":[1, true]}' \
        http://localhost:8545
{"jsonrpc":"2.0","id":1337,"error":{"code":-32602,"message":"invalid argument 0: hex string without 0x prefix"}}

~ ❯ curl -H "Content-Type: application/json" -X POST --data \
        '{"id":1337,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x1", true]}' \
        http://localhost:8545
{"jsonrpc":"2.0","id":1337,"result":null}
```

**Comment by wolflo** [comment]
My thought would be that matching geth for non-test methods is important to prevent divergence between an action's behavior against hardhat vs. mainnet. This also has the benefit of not requiring changes.

For methods that don't exist on mainnet (e.g. `evm_increaseTime`), being more permissive probably isn't bad. It seems to make sense to _either_ keep the same criteria for integer arguments as mainnet methods (meaning accept hex-encoded and reject decimal-encoded) _or_ be more permissive (accepting both). Since the first option would probably be a breaking change, allowing hex-encoded parameters in addition to the currently accepted decimal-encoded parameters seems to make the most sense.

**Comment by alcuadrado** [comment]
> My thought would be that matching geth for non-test methods is important to prevent divergence between an action's behavior against hardhat vs. mainnet. This also has the benefit of not requiring changes.

Exactly

> I think it makes sense to be more permissive and accept hex-encoded arguments in these methods. But we can't change the output of the methods since that would be a breaking change.

Agreed

> For methods that don't exist on mainnet (e.g. `evm_increaseTime`), being more permissive probably isn't bad. It seems to make sense to _either_ keep the same criteria for integer arguments as mainnet methods (meaning accept hex-encoded and reject decimal-encoded) _or_ be more permissive (accepting both). Since the first option would probably be a breaking change, allowing hex-encoded parameters in addition to the currently accepted decimal-encoded parameters seems to make the most sense.

I agree with this. the `evm` module has always felt weird, and we've had this discussion before. I believe at the time ganache didn't accept hex values, or we incorrectly assumed that.

We should accept numbers or hex values in the evm module. But for now, I'd keep the `hardhat` module as strict as it is now.

**Comment by fvictorio** [comment]
Closed by #1735.
